src/turbomind/kernels/attention/rotary_embedding.h

```diff
@@ -110,6 +116,15 @@ struct FastRoPE {
                 inv_freq_[i / 2] = (1 - smooth) * freq * llama3_inv_scaling_factor + smooth * freq;
             }
         }
+        if (yarn_ramp_max) {
+            PRAGMA_UNROLL
+            for (int i = 0; i < N; i += 2) {
+                auto  freq       = inv_freq_[i / 2];
+                float alpha      = ((idx + i) / 2 - yarn_ramp_min) / (yarn_ramp_max - yarn_ramp_min);
+                alpha            = fmaxf(0.f, fminf(1.f, alpha));
+                inv_freq_[i / 2] = freq * (1 - alpha + alpha / factor);
```

Comment:
All these expensive divisions can be done in host code. See how the llama3 rope is implemented just a few lines above.